x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapper.java

```diff
@@ -0,0 +1,519 @@
+                if (ignoreMalformed.value() && parser.currentToken().isValue()) {
+                    context.addIgnoredField(mappedFieldType.name());
+                    return;
+                } else {
+                    throw e;
+                }
+            }
+        }
+        if (numericValue == null) {
+            numericValue = nullValueNumeric;
+            if (numericValue == null) return;
+        } else {
+            numericValue = convertToSignedLong(numericValue);
+        }
+
+        boolean docValued = fieldType().hasDocValues();
+        boolean indexed = fieldType().isSearchable();
+        boolean stored = fieldType.stored();
+
+        List<Field> fields = NumberFieldMapper.NumberType.LONG.createFields(fieldType().name(), numericValue, indexed, docValued, stored);
+        context.doc().addAll(fields);
+        if (docValued == false && (indexed || stored)) {
+            createFieldNamesField(context);
+        }
+    }
+
+    @Override
+    protected void mergeOptions(FieldMapper other, List<String> conflicts) {
+        UnsignedLongFieldMapper mergeWith = (UnsignedLongFieldMapper) other;
+        if (mergeWith.ignoreMalformed.explicit()) {
+            this.ignoreMalformed = mergeWith.ignoreMalformed;
+        }
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+
+        if (includeDefaults || ignoreMalformed.explicit()) {
+            builder.field("ignore_malformed", ignoreMalformed.value());
+        }
+        if (nullValue != null) {
+            builder.field("null_value", nullValue);
+        }
+    }
+
+    /**
+     * Parse object to unsigned long
+     * @param value must represent an unsigned long in rage [0;18446744073709551615] or an exception will be thrown
+     */
+    private static long parseUnsignedLong(Object value) {
+        if ((value instanceof Long) || (value instanceof Integer) || (value instanceof Short) || (value instanceof Byte)) {
+            long lv = ((Number) value).longValue();
+            if (lv < 0) {
+                throw new IllegalArgumentException("Value [" + lv + "] is out of range for unsigned long.");
+            }
+            return lv;
+        }
+        String stringValue = (value instanceof BytesRef) ? ((BytesRef) value).utf8ToString() : value.toString();
+        try {
+            return Long.parseUnsignedLong(stringValue);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("For input string: \"" + stringValue + "\"");
+        }
+    }
+
+    /**
+     * Convert an unsigned long to the singed long by subtract 2^63 from it
+     * @param value – unsigned long value in the range [0; 2^64-1], values greater than 2^63-1 are negative
+     * @return signed long value in the range [-2^63; 2^63-1]
+     */
+    private static long convertToSignedLong(long value) {
```

Comment:
without context I'd guess that this is doing a cast, maybe make the method name more explicit, e.g. `unsignedToSortableSignedLong`